### DISCUSSION TITLE
개발 작업 기술 검토 및 규칙

### DISCUSSION THREAD

**Discussion by aespablo** [envybros/nomads-planet#5] [Announcements]
유니티 개발에 있어서 코딩에 대한 규칙과 기술 아이디어 및 검토는 필수적이다.

여기서 그 모든 것에 대해 이야기를 다룰 것이다.

**Comment by aespablo** [comment]
### 아키텍처와 폴더 구조

1. Assets: Unity 프로젝트의 모든 리소스를 저장하는 폴더
   - _Scenes: 게임 내의 각 화면과 관련된 파일들을 저장
   - _Prefabs: 여러 오브젝트들의 프리팹을 저장
   - _Scripts: 모든 스크립트 파일들을 저장
     - Managers: 게임 매니저, 사운드 매니저, VR 매니저 등의 관리 스크립트를 저장
     - UI: 사용자 인터페이스 관련 스크립트를 저장
     - Interactions: 오브젝트 상호작용 및 이벤트 처리 스크립트를 저장
   - Materials: 오브젝트의 소재 파일들을 저장
   - Textures: 텍스쳐 이미지 파일들을 저장
   - Sounds: 오디오 클립 및 배경음악 파일을 저장
   - Plugins: 외부 개발자들이 제공하는 플러그인이나 라이브러리를 저장
   - Animations: 애니메이션 클립을 저장
  
2. ProjectSettings: Unity 프로젝트 설정 파일을 저장하는 폴더

3. Packages: 프로젝트에 사용되는 패키지 및 의존성 관리 파일을 저장

추가 사항이 있을 경우, 논의 후 추가하기

**Reply by aespablo** [reply]
캐릭터 FBX 파일 내용들을 위한 Models 폴더를 하나 만들고, 내부에 각 캐릭터 파일 이름별 폴더를 만들어주면 된다.

**Comment by pulppixel** [comment]
## 롤러코스터 구현하기

[Curvy](https://assetstore.unity.com/packages/tools/utilities/curvy-splines-8-212532) 이 에셋 써서 경로 구축하면 될 것 같다. 

기타 사물들은 따로 구해보자. 

에셋 찾다보니 현실이 좀 느껴진다. 

그래도 끝까지 찾아보자잉

**Comment by aespablo** [comment]
사용할 외부 에셋 정리

# 구현에 사용할 요소

1. Game Creator 2 - 캐릭터 및 개발 지원
  - 캐릭터 데이터 저장은?
  - 파이어 베이스 검토
  - 애니메이션 등등 적용해야지

2. Curvy Splines 8 - 롤러 코스터 구현용
3. Curved World - 원더월드 행성 구현용

# 디자인 구성 요소

1. KEKOS (Cartoon Kids) - 주인공 캐릭터
2. Quirky Series - 동물팩
3. Dreamscape Nature - 자연 환경 팩
4. Space Graphics Planets - 우주에서 날아올 때 보여줄 용도? 롤러 코스터에서도 쓰일듯
5. Corals - 바다 구현할 때 사용할 거?
6. RPG Fantasy Monster Bundle #6 - 괴물이나 NPC 구현하기 용도

- Flat Kit: Toon Shading and Water - 전체적으로 툰 셰이딩 적용하면 어떨지..?!

# VFX

1. 100 Special Skills Effects Pack
2. Ultimate VFX

**Reply by aespablo** [reply]
# 워터 셰이더 관련해서,,

Weather Maker - Volumetric Clouds and Weather System for Unity

이게 더 좋다네?

**Reply by aespablo** [reply]
# VFX 추가

MK Glow - Bloom & Lens & Glare

**Reply by aespablo** [reply]
# VR 개발 지원

VR Interaction Framework

드디어 이거 써보냐 ㅋㅋㅋ

**Comment by aespablo** [comment]
근데 유료 에셋을 저기다가 두기 좀 그래...
플러그인을 아예 Gitignore로 해두고,, USB에 프로젝트 넣어다니는 건 어때?

**Comment by aespablo** [comment]
자동차 같은 시뮬레이션 작업들은 DOTS 이용해보기

https://youtu.be/H7zAORa3Ux0

Unity의 DOTS (Data-Oriented Technology Stack)는 고성능 멀티스레딩 코드를 작성하기 위한 Unity의 새로운 기술 스택입니다. 이는 세 가지 주요 기술로 구성되어 있습니다: ECS (Entity Component System), Job System, 그리고 Burst Compiler.

1. **ECS (Entity Component System)**: ECS는 새로운 프로그래밍 패러다임으로, 데이터를 중심으로 시스템을 설계합니다. 기존의 객체 지향 설계 방식이 "데이터와 기능이 결합된 객체"를 중심으로 하는 데 비해, ECS는 데이터와 로직을 분리하여 관리합니다. 이를 통해 코드를 보다 효율적으로 작성하고, 성능을 향상시킬 수 있습니다.

2. **Job System**: 이는 위에서 설명한 것처럼, Unity의 멀티스레딩 솔루션입니다. Job System을 사용하면 CPU 작업을 여러 개의 작은 작업으로 나누고 이를 병렬로 실행할 수 있습니다. 이렇게 하면 많은 양의 계산을 빠르게 처리할 수 있습니다.

3. **Burst Compiler**: Burst Compiler는 Unity의 새로운 LLVM 기반 컴파일러로, 고성능 코드를 생성할 수 있습니다. Burst Compiler는 Job System과 함께 작동하여, 일반적인 C# 코드를 빠르게 실행되는 저수준 코드로 변환합니다.

이 세 가지 기술을 함께 사용하면, 데이터 지향적인 설계와 고성능 멀티스레딩을 통해 게임의 성능을 크게 향상시킬 수 있습니다. Unity의 DOTS는 특히 CPU 집약적인 작업, 예를 들어 큰 규모의 AI 시뮬레이션, 물리 시뮬레이션, 복잡한 게임 로직 등에서 큰 장점을 보입니다.

**Comment by aespablo** [comment]
Curved World의 Planet Y <- 이거 그대로 쓰면 돼.. 맵은 간단하게 가자

행성 사이에 다리하나 두고, 한강 구현하면 더 좋을듯

**Reply by aespablo** [reply]
이거 없드라..

**Comment by aespablo** [comment]
바닥 타일들 15x15 이런거 전부 5x5로 변경 (그래야 더 유연하게 휘어짐)

**Comment by pulppixel** [comment]
이거 어차피 다 gitignore해놔서 큰 문제 없어보이는데

issue따서 브랜치 만들거 작업해도 괜찮을 것 같네
